### PR TITLE
Add support for `append` mode for Azure Blob Storage

### LINKED ISSUE #836: [azure] Support for "append" mode for Azure Blobs
Hi all!

I use smart-open for one of my projects and I've recently run into the need for "append" mode for Azure blobs. This is something Azure's SDK supports natively but it looks like it hasn't been implemented in smart-open yet. 

I was thinking on adding support for this feature myself but I was wondering if there is any additional concern/inconvenience I might be missing. 

P.S.: Thanks for such a simple yet useful tool!

Cheers.

### PR DESCRIPTION
**Pull request by geovalexis**
### Motivation

Append mode is not a widely used feature in Azure Blob Storage, but it can be particularly useful for specific scenarios, such as writing logs or updating sentinel or marker files. As the native Azure Blob Storage SDK provides support for append mode, this PR integrates that functionality into the `smart_open` interface, making it more accessible and convenient for developers.

This implementation comes with a few limitations that are worth noting:

* There is a maximum size of the blob that can be written in append mode: 195 GiB. This is a limitation of the [Azure Blob Storage API itself](https://learn.microsoft.com/en-us/rest/api/storageservices/understanding-block-blobs--append-blobs--and-page-blobs#about-append-blobs) and cannot be changed.
* The blob must be created in append mode on the first place, meaning that we cannot simply append to an existing blob that may have been created in a different mode (e.g. block or page mode).

As per these limitations, we observe that the append mode is not suitable for all use cases and is not meant to be used indiscriminately by anyone looking to append data to a blob. However, bearing them in mind, many developers may find it convenient for their use cases, especially when they are already using `smart_open` for other operations.

Closes #836

### Tests

New tests have been added to cover the new functionality, including:
- Writing to a new blob in append mode
- Appending to an existing blob in append mode
- Attempting to append to a blob that was not created in append mode (which should raise an error)

Also, I took the chance to fix a few existing tests that were failing when using Azurite. The method `BlobClient._stage_contents` does not exist in the Azure Blob Storage SDK (or at least not in the latest versions), so I replaced it with `BlobClient.get_block_list("uncommitted")`, which returns the same information. 

### Checklist

- [X] Picked a concise, informative and complete title
- [X] Clearly explained the motivation behind the PR
- [X] Linked to any existing issues that your PR will be solving
- [X] Included tests for any new functionality
- [X] Checked that all unit tests pass

### CODE REVIEW

File: smart_open/azure.py [hunk 1]
```diff
@@ -550,3 +558,94 @@ def __repr__(self):
...
+    def seek(self, offset, whence=smart_open.constants.WHENCE_START):
+        """Unsupported."""
+        raise io.UnsupportedOperation
+
+    def truncate(self, size=None):
+        """Unsupported."""
+        raise io.UnsupportedOperation
+
+    def detach(self):
+        raise io.UnsupportedOperation("detach() not supported")
```
[the remark this answers is not in the document]
Comment (ddelange) [reply smart_open/azure.py:613]: > why is this raise different from the others?

aha, I see it's copy pasted from above. let's leave it then.

File: smart_open/azure.py [hunk 2]
```diff
@@ -550,3 +558,94 @@ def __repr__(self):
...
+            )
+
+        # Uploads data as an AppendBlob type with automatic block chunking.
+        # The AppendBlob will be created at first if it does not exist or append to it if it does already.
+        return self._blob.upload_blob(
+            data=b,
+            blob_type=azure.storage.blob.BlobType.APPENDBLOB,
+            overwrite=False,
+            **self._blob_kwargs,
+        )
```
Comment (ddelange) [inline smart_open/azure.py:628]: can you implement the `_min_part_size` buffer mechanic similar to the azure `Writer` class? this current implementation is unbuffered, and you'll quickly hit the max parts limit an AppendBlob support if you do many small writes

File: tests/test_azure.py [hunk 3]
```diff
@@ -856,3 +862,138 @@ def test_flush_close(self):
...
+
+        with smart_open.azure.AppendWriter(CONTAINER_NAME, blob_name, CLIENT) as fout:
+            with gzip.GzipFile(fileobj=fout, mode='w') as zipfile:
+                zipfile.write(expected)
+
+        with smart_open.azure.Reader(CONTAINER_NAME, blob_name, CLIENT) as fin:
+            with gzip.GzipFile(fileobj=fin) as zipfile:
+                actual = zipfile.read()
+
+        self.assertEqual(expected, actual)
```
[the remark this answers is not in the document]
Comment (geovalexis) [reply tests/test_azure.py]: addressed 👍

File: smart_open/azure.py (hunk 2 above)
Comment (geovalexis) [reply smart_open/azure.py]: addressed 👍

**Comment by geovalexis** [comment]
hey! Sorry for the delay on this! it completely fell off my radar, I'm trying to resume my work on this and hopefully get it merged at some point.

**Comment by ddelange** [comment]
@geovalexis is this ready for review?

**Comment by geovalexis** [comment]
> @geovalexis is this ready for review?

Yes, it is

File: tests/test_azure.py [hunk 4]
```diff
@@ -856,3 +887,141 @@ def test_flush_close(self):
...
+        test_string = u"žluťoučký koníček".encode('utf8')
+        blob_name = "test_append_existing_write_blob_%s" % BLOB_NAME
+
+        # Creating blob of type BlockBlob
+        with smart_open.azure.Writer(CONTAINER_NAME, blob_name, CLIENT) as fout:
+            fout.write(test_string)
+
+        with self.assertRaises(
+            azure.core.exceptions.ResourceExistsError, msg="The blob type is invalid for this operation."
+        ):
```
Comment (ddelange) [inline tests/test_azure.py:945]: noting here that this error is thrown upon write() not upon open(). i think that's analogous to for instance s3 read() on a non-existent key

File: tests/test_azure.py [hunk 5]
```diff
@@ -856,3 +887,141 @@ def test_flush_close(self):
...
+            "azure://%s/%s" % (CONTAINER_NAME, blob_name),
+            "rb",
+            transport_params=dict(client=CLIENT),
+        ))
+        self.assertEqual(output, [test_string_1 + test_string_2 + test_string_3])
+
+    def test_append_block_over_max_block_size(self):
+        """
+        Does appending into an Azure Blob file work correctly when the block size is over the max block size?
+        By default, this block size is 4MB. Refer to official Azure documentation for more information:
```
Comment (ddelange) [inline tests/test_azure.py:995]: conversely, could you add a characterization test for the min_part_size mechanic? ie first write of min_part_size - 1 doesnt trigger upload, then writing 1 byte triggers upload? 

_DEFAULT_MIN_PART_SIZE is set to 64MB in azure.py so that edge case is already implicitly covered. make it explicit? some way we can hit this 4MB limit or is it completely abstracted away for the smart_open end-user?

**Review by ddelange — COMMENTED** [review]
i like the test coverage! can you add a note to the end of [this](https://github.com/piskvorky/smart_open?tab=readme-ov-file#azure-advanced-usage) README section mentioning mode="ab"? no example needed, just mention the functionality?

File: tests/test_azure.py (hunk 5 above)
Comment (geovalexis) [reply tests/test_azure.py:995]: Updated code with new requested unit test 👍 . About 4Mb limit, that's completely abstracted away from smart-open, Azure SDK's handled this automatically by splitting the data into 4Mb chunks if necessary.

**Comment by geovalexis** [comment]
@ddelange thanks for reviewing my PR and your comments. I've pushed new changes addressing your comments, please do not hesitate to push back if there's something I missed.

File: README.rst [hunk 6]
```diff
@@ -467,6 +467,8 @@ Additional keyword arguments can be propagated to the ``commit_block_list`` meth
     kwargs = {'metadata': {'version': 2}}
     fout = open('azure://container/key', 'wb', transport_params={'blob_kwargs': kwargs})
 
+``smart_open`` also supports append mode (``mode="ab"``) for Azure Blob Storage using `Append Blobs <https://learn.microsoft.com/en-us/rest/api/storageservices/understanding-block-blobs--append-blobs--and-page-blobs#about-append-blobs>`__. Note that the blob must have been created as an Append Blob and is subject to a maximum size of 195 GiB.
```
Comment (ddelange) [inline README.rst]:
```suggestion
For Azure, ``smart_open`` also supports append mode (``mode="ab"``) by creating new (or appending to existing) `Append Blobs <https://learn.microsoft.com/en-us/rest/api/storageservices/understanding-block-blobs--append-blobs--and-page-blobs#about-append-blobs>`__ (max 195 GiB).
```

**Review by ddelange — APPROVED** [review]
@geovalexis many thanks for your contribution!
